common/src/main/java/org/apache/celeborn/common/network/server/DataPartitionReader.java

```diff
@@ -285,47 +284,56 @@ private int readBuffer(
 
   private void updateConsumingOffset() throws IOException {
     while (currentPartitionRemainingBytes == 0
-        && (currentDataRegion < numRegions - 1 || numRemainingPartitions > 0)) {
-      if (numRemainingPartitions <= 0) {
+        && (currentDataRegion < numRegions - 1 || numRemainingPartitionsInCurrentRegion > 0)) {
+      if (numRemainingPartitionsInCurrentRegion <= 0) {
         ++currentDataRegion;
-        numRemainingPartitions = endPartitionIndex - startPartitionIndex + 1;
+        numRemainingPartitionsInCurrentRegion = endPartitionIndex - startPartitionIndex + 1;
 
         // read the target index entry to the target index buffer
         indexFileChannel.position(
             currentDataRegion * getIndexRegionSize()
                 + (long) startPartitionIndex * INDEX_ENTRY_SIZE);
-        readHeaderOrIndexBuffer(indexFileChannel, indexBuffer, indexBuffer.capacity());
+        readFully(indexFileChannel, indexBuffer, indexBuffer.capacity());
       }
 
       // get the data file offset and the data size
       dataConsumingOffset = indexBuffer.getLong();
       currentPartitionRemainingBytes = indexBuffer.getLong();
-      --numRemainingPartitions;
+      --numRemainingPartitionsInCurrentRegion;
 
       logger.debug(
           "readBuffer updateConsumingOffset, {},  {}, {}, {}",
           streamId,
-          dataFileChannel.size(),
+          dataFileChannelWithPosition.channel.size(),
           dataConsumingOffset,
           currentPartitionRemainingBytes);
 
       // if these checks fail, the partition file must be corrupted
       if (dataConsumingOffset < 0
-          || dataConsumingOffset + currentPartitionRemainingBytes > dataFileChannel.size()
+          || dataConsumingOffset + currentPartitionRemainingBytes
+              > dataFileChannelWithPosition.channel.size()
           || currentPartitionRemainingBytes < 0) {
         throw new RuntimeException("File " + fileInfo.getFilePath() + " is corrupted");
       }
     }
   }
 
+  /**
+   * @param buffer
+   * @return true if current region has more data after reading
+   * @throws IOException
+   */
   private boolean readBuffer(ByteBuf buffer) throws IOException {
     try {
-      dataFileChannel.position(dataConsumingOffset);
+      if (dataFileChannelWithPosition.position != dataConsumingOffset) {
```

Comment:
we can make dataFileChannelWithPosition provide a position method to do this.

Reply:
done